README.md
# ProAut Parameter

## Introduction

This is a simple helper-package for loading variables from the parameter
server.

The main idea is to simplify the handling of parameters, which can be set
to default values. The _load()_ function will print if a parameter is
loaded from the server or if its default value is used. In each case, the
final value is printed, too.


## Node

```
rosrun parameter_pa parameter_pa_node
roslaunch parameter_pa parameter_pa.launch
```


### Services:

Service Name            | Type                                                                                                                | Description
------------------------|---------------------------------------------------------------------------------------------------------------------|---------------------------------
"~/substitutePath"      | [parameter_pa/ParameterPaString](https://github.com/TUC-ProAut/ros_parameter/blob/master/srv/ParameterPaString.srv) | Replaces each "$(find xyz)" with the corresponding path - this is similar to calling _loadPath_().
"~/substituteRessource" | [parameter_pa/ParameterPaString](https://github.com/TUC-ProAut/ros_parameter/blob/master/srv/ParameterPaString.srv) | Resolves ressource name - this is similar to calling _loadTopic_().


## Functions

The main _load()_ function can handle bool, int, double and string - all as
single values and as vectors. Also matrices of double (float) can be loaded.

Additionaly, there are two more dedicated load functions: _loadPath()_
and _loadTopic()_. Both are basically loading a simple string from the
parameter server as descript above. Furthermore the strings are specificly
manipulated:

### loadPath()
Every "$(find xyz)" will be replaced with the path of package xyz.

### loadTopic()
Names are resolved with additional operators.

For these examples assume that ROS\_NAMESPACE is set to "/name/space" and
that the nodes' name is "parameter_pa_node"

#### Regular ressource names

in_string      | out_string     | comment
---------------|----------------|---------
ressource      | ressource      | relative name
ressource/     | ressource/     | relative name ending with slash
ns/ressource   | ns/ressource   | relative name with namespace
ns/ressource/  | ns/ressource/  | relative name with namespace ending with slash
/ressource     | /ressource     | absolute name
/ressource/    | /ressource/    | absolute name ending with slash
/ns/ressource  | /ns/ressource  | absolute name with namespace
/ns/ressource/ | /ns/ressource/ | absolute name with namespace ending with slash

#### Private ressource names

in_string       | out_string                                    | comment
----------------|-----------------------------------------------|---------
~               | /name/space/parameter\_pa\_node               | private namespace
~/              | /name/space/parameter\_pa\_node/              | private namespace ending with slash
~/ressource     | /name/space/parameter\_pa\_node/ressource     | private namespace with relative name
~/ns/ressource/ | /name/space/parameter\_pa\_node/ns/ressource/ | private namespace with relative name and relative namespace ending in slash
/abc/~/xyz      | Error                                         | no private namespace within absolute name allowed

#### Locale ressource names

in_string       | out_string                | comment
----------------|---------------------------|---------
.               | /name/space               | local namespace
./              | /name/space/              | local namespace ending with slash
./ressource/    | /name/space/ressource/    | local namespace with relative name ending in slash
./ns/ressource  | /name/space/ns/ressource  | local namespace with relative name and relative namespace
/abc/./xyz      | /abc/xyz                  | local namespace within absolute name

#### Previous ressource names

in_string       | out_string                | comment
----------------|---------------------------|---------
..              | /name                     | previous namespace
../             | /name/                    | previous namespace ending with slash
../ressource/   | /name/ressource/          | previous namespace with relative name ending in slash
../ns/ressource | /name/ns/ressource        | previous namespace with relative name and relative namespace
/abc/def/../xyz | /abc/xyz                  | previous namespace within absolute name

#### Removal of double slashes

in_string      | out_string     | comment
---------------|----------------|---------
ressource//    | ressource/     | relative name ending with slash
//ns/ressource | /ns/ressource  | absolute name with namespace

#### More examples

in_string                         | out_string
----------------------------------|------------
~/abc/.//def/../ns//ressource/    | /name/space/parameter\_pa\_node/abc/ns/ressource/
ns//.././..//ressource            | /name/space/ressource
../../../../../../ressource/      | /ressource/


## Example Usage

A typical output looks like this:
```
[ INFO] [1499357951.543886548]: load parameter ~/tf_lookup_time (0.1)
[ INFO] [1499357951.544331412]: parameter ~/buffer_pointcloud not set (defaults to false)
[ INFO] [1499357951.544867951]: load parameter ~/debugging (true)
```

Here is the related source code, which is taken from [pcdfilter_pa](https://github.com/TUC-ProAut/ros_pcdfilter/blob/master/src/pcdfilter_pa_node.cpp):
```
cParameterPaRos paramloader;

// ...
paramloader.load("~/tf_lookup_time"   , filters_.tf_lookup_time_     );
paramloader.load("~/buffer_pointcloud", rosparams_.buffer_pointcloud_);
paramloader.load("~/debugging"        , rosparams_.debugging_        );
```


## Links and packages

### Source code at github
https://github.com/TUC-ProAut/ros_parameter

### ROS packages
ros-kinetic-parameter-pa

ros-melodic-parameter-pa _(upcoming)_

ros-noetic-parameter-pa _(upcoming)_

## ROS Build-Status and Documentation

ROS-Distribution | Build-Status                                                                                                                                                            | Documentation
-----------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------
Indigo           | EOL April 2019                                                                                                                                                          | [docs.ros.org/indigo](http://docs.ros.org/indigo/api/parameter_pa/html/index.html)
Jade             | EOL May 2017                                                                                                                                                            | [docs.ros.org/jade](http://docs.ros.org/jade/api/parameter_pa/html/index.html)
Kinetic          | EOL April 2021                                                                                                                                                          | [docs.ros.org/kinetic](http://docs.ros.org/kinetic/api/parameter_pa/html/index.html)
Lunar            | EOL May 2019                                                                                                                                                            | [docs.ros.org/lunar](http://docs.ros.org/lunar/api/parameter_pa/html/index.html)
Melodic          | [![Build Status](http://build.ros.org/buildStatus/icon?job=Mdev__parameter_pa__ubuntu_bionic_amd64)](http://build.ros.org/job/Mdev__parameter_pa__ubuntu_bionic_amd64/) | [docs.ros.org/melodic](http://docs.ros.org/melodic/api/parameter_pa/html/index.html)
Noetic           | [![Build Status](http://build.ros.org/buildStatus/icon?job=Ndev__parameter_pa__ubuntu_focal_amd64)](http://build.ros.org/job/Ndev__parameter_pa__ubuntu_focal_amd64/) | [docs.ros.org/noetic](http://docs.ros.org/noetic/api/parameter_pa/html/index.html)


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package parameter_pa
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.2.3 (2021-06-08)
------------------
* bumped cmake Version to 3.0.2 to avoid CMP0048
* Contributors: Peter Weissig

1.2.2 (2021-06-08)
------------------
* updated documentation
  + updated README.md
  + added rosdoc.yaml
  + removed doxygen.h
* updated year of copyright
* updated gitignore-file
* updated files to new repository (only changed documentation and license)
  moved from github.com/peterweissig/ros_parameter to github.com/TUC-ProAut/ros_parameter
* Contributors: Peter Weissig

1.2.1 (2017-12-14)
------------------
* bugfixed loadTopic (loaded topic was not resolved)
* bugfixed documentation (some typos)
* Contributors: Peter Weissig

1.2.0 (2017-12-11)
------------------
* removed unnecessary headers
* bugfixed include of xmlrpcpp
  (before kinetic those files were not related to a package)
* update README.md
  bugfixed formatting (missing blanc line before table)
* added new features ("./" and "../" within ressource names)
  added visual studio code config files (for ros kinetic)
  updated documentation
* Contributors: Peter Weissig

1.1.0 (2017-08-01)
------------------
* moved header from include/ to include/${project_name}
  also fixed related paths
* Contributors: Peter Weissig

1.0.2 (2017-07-31)
------------------
* updated year within license
* updated version number
* added install target to CMakeLists.txt
* Contributors: Peter Weissig

1.0.0
-----
* updated package.xml to format version 2
* updated doxygen.h and README.md
* updated CMakeLists.txt and package.xml
  added Eigen3
* updated git policies
* Initial commit
* Contributors: Peter Weissig


checkout.sh
#!/bin/sh

###############################################################################
#                                                                             #
# checkout.sh                                                                 #
# ===========                                                                 #
#                                                                             #
###############################################################################
#                                                                             #
# github repository                                                           #
#   https://github.com/TUC-ProAut/ros_parameter                               #
#                                                                             #
# Chair of Automation Technology, Technische Universität Chemnitz             #
#   https://www.tu-chemnitz.de/etit/proaut                                    #
#                                                                             #
###############################################################################
#                                                                             #
# New BSD License                                                             #
#                                                                             #
# Copyright (c) 2015-2021, Peter Weissig, Technische Universität Chemnitz     #
# All rights reserved.                                                        #
#                                                                             #
# Redistribution and use in source and binary forms, with or without          #
# modification, are permitted provided that the following conditions are met: #
#     * Redistributions of source code must retain the above copyright        #
#       notice, this list of conditions and the following disclaimer.         #
#     * Redistributions in binary form must reproduce the above copyright     #
#       notice, this list of conditions and the following disclaimer in the   #
#       documentation and/or other materials provided with the distribution.  #
#     * Neither the name of the Technische Universität Chemnitz nor the       #
#       names of its contributors may be used to endorse or promote products  #
#       derived from this software without specific prior written permission. #
#                                                                             #
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" #
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE   #
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE  #
# ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY      #
# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES  #
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR          #
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER  #
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT          #
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY   #
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH #
# DAMAGE.                                                                     #
#                                                                             #
###############################################################################

NAME_GIT_THIS="parameter"
NAME_GIT_SUBS=""

###############################################################################
URL_GIT_BASE="https://github.com/TUC-ProAut/"
URL_GIT_THIS="${URL_GIT_BASE}ros_${NAME_GIT_THIS}.git"
DIR_THIS="${NAME_GIT_THIS}_pa/"

NAME_CHECKOUT_SCRIPT="checkout.sh"
NAME_ARGUMENT_SCRIPT="dependency"

###############################################################################
if [ "$#" -eq 0 ]; then
    echo "Running this script without parameters."
    echo "Therefore the project"
    echo "  \"${NAME_GIT_THIS}\""
    echo "will be checked out completely."
    echo ""

    SRC_DIR_MAIN="$(pwd)/"

    if [ -d ".git" ]; then
        echo "This folder already is a git-repository!"
        return
    fi
    git clone "${URL_GIT_THIS}" "${NAME_GIT_THIS}"
    if [ $? -ne 0 ]; then return; fi

    echo ""
    echo "### checking out additional source code"
    cd "${NAME_GIT_THIS}"
    if [ ! -f "${NAME_CHECKOUT_SCRIPT}" ]; then
        echo "Error - no checkout script"
        return
    fi
    sh "${NAME_CHECKOUT_SCRIPT}" "${NAME_ARGUMENT_SCRIPT}"
    if [ $? -ne 0 ]; then return; fi

    echo ""
    echo "### deleting this script"
    cd "${SRC_DIR_MAIN}"
    rm "${NAME_CHECKOUT_SCRIPT}"

###############################################################################
elif [ "$1" = "${NAME_ARGUMENT_SCRIPT}" ]; then
    echo "Running this script with argument \"$1\"."
    echo "Therefore all dependend sources of"
    echo "  \"${NAME_GIT_THIS}\""
    echo "will be checked out recursivly."
    echo ""

    SRC_DIR_DEPENDENCY="$(pwd)"

    for dependency in $NAME_GIT_SUBS ; do
        cd "${SRC_DIR_DEPENDENCY}"

        DIR_DEPENDENCY="${dependency}_pa/"

        if [ ! -d "../${DIR_DEPENDENCY}" ]; then
            echo ""
            echo "### checking out ${dependency}"
            cd ".."

            URL="${URL_GIT_BASE}ros_${dependency}.git"
            git clone "${URL}" "${DIR_DEPENDENCY}"

            cd "${DIR_DEPENDENCY}"
            if [ -f "checkout.sh" ]; then
                sh "${NAME_CHECKOUT_SCRIPT}" "${NAME_ARGUMENT_SCRIPT}"
            fi
        fi
    done

###############################################################################
else
    echo "Running this script with argument \"$1\"."
    echo "Error - expecting no argument or \"${NAME_ARGUMENT_SCRIPT}\"."
    return
fi


src/parameter_pa_node.cpp
/******************************************************************************
*                                                                             *
* parameter_pa_node.cpp                                                       *
* =====================                                                       *
*                                                                             *
*******************************************************************************
*                                                                             *
* github repository                                                           *
*   https://github.com/TUC-ProAut/ros_parameter                               *
*                                                                             *
* Chair of Automation Technology, Technische Universität Chemnitz             *
*   https://www.tu-chemnitz.de/etit/proaut                                    *
*                                                                             *
*******************************************************************************
*                                                                             *
* New BSD License                                                             *
*                                                                             *
* Copyright (c) 2015-2021, Peter Weissig, Technische Universität Chemnitz     *
* All rights reserved.                                                        *
*                                                                             *
* Redistribution and use in source and binary forms, with or without          *
* modification, are permitted provided that the following conditions are met: *
*     * Redistributions of source code must retain the above copyright        *
*       notice, this list of conditions and the following disclaimer.         *
*     * Redistributions in binary form must reproduce the above copyright     *
*       notice, this list of conditions and the following disclaimer in the   *
*       documentation and/or other materials provided with the distribution.  *
*     * Neither the name of the Technische Universität Chemnitz nor the       *
*       names of its contributors may be used to endorse or promote products  *
*       derived from this software without specific prior written permission. *
*                                                                             *
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" *
* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE   *
* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE  *
* ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY      *
* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES  *
* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR          *
* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER  *
* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT          *
* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY   *
* OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH *
* DAMAGE.                                                                     *
*                                                                             *
******************************************************************************/

// local headers
#include "parameter_pa/parameter_pa_node.h"
#include "parameter_pa/parameter_pa_ros.h"

// standard headers
#include <string>
#include <math.h>

//**************************[main]*********************************************
int main(int argc, char **argv) {

    ros::init(argc, argv, "pcd_filter_pa_node");
    cParameterPaNode pcd_filter;

    ros::spin();

    return 0;
}

//**************************[cParameterPaNode]*********************************
cParameterPaNode::cParameterPaNode() {

    cParameterPaRos paramloader;
    std::string str_service = paramloader.resolveRessourcename("~/");

    // service for path substitution
    ser_path_ = nh_.advertiseService( str_service + "substitutePath",
      &cParameterPaNode::substitutePathCallbackSrv, this);
    // service for ressource name substitution
    ser_ressource_ = nh_.advertiseService(
      str_service + "substituteRessource",
      &cParameterPaNode::substituteNameCallbackSrv, this);
}

//**************************[~cParameterPaNode]********************************
cParameterPaNode::~cParameterPaNode() {

}

//**************************[substitutePathCallbackSrv]*************************
bool cParameterPaNode::substitutePathCallbackSrv(
  parameter_pa::ParameterPaString::Request  &req,
  parameter_pa::ParameterPaString::Response &res) {

    std::string str = req.in_string;
    res.ok = cParameterPaRos::replaceFindpack(str);
    res.out_string = str;

    return true;
}

//**************************[substituteNameCallbackSrv]*************************
bool cParameterPaNode::substituteNameCallbackSrv(
  parameter_pa::ParameterPaString::Request  &req,
  parameter_pa::ParameterPaString::Response &res) {

    res.out_string = cParameterPaRos::resolveRessourcename(req.in_string);
    res.ok = true;
    
    return true;
}


include/parameter_pa/parameter_pa_ros.h
/******************************************************************************
*                                                                             *
* parameter_pa_ros.h                                                          *
* ==================                                                          *
*                                                                             *
*******************************************************************************
*                                                                             *
* github repository                                                           *
*   https://github.com/TUC-ProAut/ros_parameter                               *
*                                                                             *
* Chair of Automation Technology, Technische Universität Chemnitz             *
*   https://www.tu-chemnitz.de/etit/proaut                                    *
*                                                                             *
*******************************************************************************
*                                                                             *
* New BSD License                                                             *
*                                                                             *
* Copyright (c) 2015-2021, Peter Weissig, Technische Universität Chemnitz     *
* All rights reserved.                                                        *
*                                                                             *
* Redistribution and use in source and binary forms, with or without          *
* modification, are permitted provided that the following conditions are met: *
*     * Redistributions of source code must retain the above copyright        *
*       notice, this list of conditions and the following disclaimer.         *
*     * Redistributions in binary form must reproduce the above copyright     *
*       notice, this list of conditions and the following disclaimer in the   *
*       documentation and/or other materials provided with the distribution.  *
*     * Neither the name of the Technische Universität Chemnitz nor the       *
*       names of its contributors may be used to endorse or promote products  *
*       derived from this software without specific prior written permission. *
*                                                                             *
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" *
* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE   *
* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE  *
* ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY      *
* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES  *
* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR          *
* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER  *
* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT          *
* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY   *
* OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH *
* DAMAGE.                                                                     *
*                                                                             *
******************************************************************************/

#ifndef __PARAMETER_PA_ROS_H
#define __PARAMETER_PA_ROS_H

// ros headers
#include <ros/ros.h>

// standard headers
#include <string>
#include <vector>

#include <eigen3/Eigen/Core>

//**************************[cParameterPaRos]**********************************
class cParameterPaRos {
  public:
    bool load(const std::string name, bool        &value,
              const bool print_default = true) const;

    bool load(const std::string name, std::string &value,
              const bool print_default = true) const;
    bool loadTopic(const std::string name, std::string &value,
                   const bool print_default = true) const;
    bool loadPath(const std::string name, std::string &value,
                  const bool print_default = true) const;

    bool load(const std::string name, int         &value,
              const bool print_default = true) const;

    bool load(const std::string name, double      &value,
              const bool print_default = true) const;


    bool load(const std::string name, std::vector<bool       > &value,
              const bool print_default = true) const;

    bool load(const std::string name, std::vector<std::string> &value,
              const bool print_default = true) const;

    bool load(const std::string name, std::vector<int        > &value,
              const bool print_default = true) const;

    bool load(const std::string name, std::vector<double     > &value,
              const bool print_default = true) const;

    bool load(const std::string name, Eigen::MatrixXf &value,
              const bool print_default = true) const;

    static bool replaceFindpack(std::string &path);
    static std::string resolveRessourcename(const std::string name);
    static std::string boolToStr(const bool value);

    // deprecated function names, just for compatibility
    // (starting after next ros-release, those names will
    //  marked as deprecated)
    // __attribute__ ((deprecated))
    bool load_topic(const std::string name, std::string &value,
                    const bool print_default = true) const;
    // __attribute__ ((deprecated))
    bool load_path(const std::string name, std::string &value,
                   const bool print_default = true) const;

    //__attribute__ ((deprecated))
    static bool replace_findpack(std::string &path);
    //__attribute__ ((deprecated))
    static void resolve_ressourcename(std::string &name);
    //__attribute__ ((deprecated))
    static std::string bool_to_str(const bool value);

  private:
    void loadSub(const std::string &n, const std::string &v,
                 const bool p, const bool r) const;
    static std::list<std::string> splitRessourcename(
      const std::string name);

    ros::NodeHandle nh_;
};

#endif // __PARAMETER_PA_ROS_H


doxygen.h
/** @mainpage ProAut Parameter
 *
 * @section intro_sec Introduction
 *
 * This is a simple helper-package for loading variables from the parameter
 * server.
 *
 * The main idea is to simplify the handling of parameters, which can be set to
 * default values. The <em>load()</em> function will print if a parameter is loaded from
 * the server or if its default value is used. In each case, the final value
 * is printed, too.
 *
 *
 * @section node_sec Node
 * @verbatim
rosrun parameter_pa parameter_pa_node
roslaunch parameter_pa parameter_pa.launch
@endverbatim
 *
 * <b>Services:</b>
 *
 * Service Name            | Type                                                                                                                           | Description
 * ------------------------|--------------------------------------------------------------------------------------------------------------------------------|---------------------------------
 * "~/substitutePath"      | <a href="https://github.com/TUC-ProAut/ros_parameter/blob/master/srv/ParameterPaString.srv">parameter_pa/ParameterPaString</a> | Replaces each "$(find xyz)" with the corresponding path - this is similar to calling <em>loadPath()</em>.
 * "~/substituteRessource" | <a href="https://github.com/TUC-ProAut/ros_parameter/blob/master/srv/ParameterPaString.srv">parameter_pa/ParameterPaString</a> | Resolves ressource name - this is similar to calling <em>loadTopic()</em>.
 *
 *
 * @section func_sec Functions
 *
 * The main <em>load()</em> function can handle bool, int, double and string - all as
 * single values and as vectors. Also matrices of double (float) can be loaded.
 *
 * Additionaly, there are two more dedicated load functions: <em>loadPath()</em> and
 * <em>loadTopic()</em>. Both are basically loading a simple string from the parameter
 * server as descript above. Furthermore the strings are specificly manipulated:
 *
 * @b loadPath() \n
 * Every "$(find xyz)" will be replaced with the path of package xyz.
 *
 * @b loadTopic() \n
 * Names are resolved with additional operators.
 *
 * For these examples assume that ROS_NAMESPACE is set to "/name/space" and that the nodes' name is "parameter_pa_node"
 *
 * Regular ressource names (<em>no change</em>):
 * in_string      | out_string     | comment
 * ---------------|----------------|--------
 * ressource      | ressource      | relative name
 * ressource/     | ressource/     | relative name ending with slash
 * ns/ressource   | ns/ressource   | relative name with namespace
 * ns/ressource/  | ns/ressource/  | relative name with namespace ending with slash
 * /ressource     | /ressource     | absolute name
 * /ressource/    | /ressource/    | absolute name ending with slash
 * /ns/ressource  | /ns/ressource  | absolute name with namespace
 * /ns/ressource/ | /ns/ressource/ | absolute name with namespace ending with slash
 *
 * Private ressource names (<em>~/</em>):
 * in_string       | out_string                                  | comment
 * ----------------|---------------------------------------------|--------
 * ~               | /name/space/parameter_pa_node               | private namespace
 * ~/              | /name/space/parameter_pa_node/              | private namespace ending with slash
 * ~/ressource     | /name/space/parameter_pa_node/ressource     | private namespace with relative name
 * ~/ns/ressource/ | /name/space/parameter_pa_node/ns/ressource/ | private namespace with relative name and relative namespace ending in slash
 * /abc/~/xyz      | Error                                       | no private namespace within absolute name allowed
 *
 * Locale ressource names (<em>./</em>):
 * in_string       | out_string                | comment
 * ----------------|---------------------------|--------
 * .               | /name/space               | local namespace
 * ./              | /name/space/              | local namespace ending with slash
 * ./ressource/    | /name/space/ressource/    | local namespace with relative name ending in slash
 * ./ns/ressource  | /name/space/ns/ressource  | local namespace with relative name and relative namespace
 * /abc/./xyz      | /abc/xyz                  | local namespace within absolute name
 *
 * Previous ressource names (<em>../</em>):
 * in_string       | out_string                | comment
 * ----------------|---------------------------|--------
 * ..              | /name                     | previous namespace
 * ../             | /name/                    | previous namespace ending with slash
 * ../ressource/   | /name/ressource/          | previous namespace with relative name ending in slash
 * ../ns/ressource | /name/ns/ressource        | previous namespace with relative name and relative namespace
 * /abc/def/../xyz | /abc/xyz                  | previous namespace within absolute name
 *
 * Removal of double slashes:
 * in_string      | out_string     | comment
 * ---------------|----------------|--------
 * ressource//    | ressource/     | relative name ending with slash
 * //ns/ressource | /ns/ressource  | absolute name with namespace
 *
 * More examples:
 * in_string                         | out_string
 * ----------------------------------|-----------
 * ~/abc/.//def/../ns//ressource/    | /name/space/parameter_pa_node/abc/ns/ressource/
 * ns//.././..//ressource            | /name/space/ressource
 * ../../../../../../ressource/      | /ressource/
 *
 *
 * @section example_sec Example
 *
 * A typical output looks like this:
 * @verbatim
[ INFO] [1499357951.543886548]: load parameter ~/tf_lookup_time (0.1)
[ INFO] [1499357951.544331412]: parameter ~/buffer_pointcloud not set (defaults to false)
[ INFO] [1499357951.544867951]: load parameter ~/debugging (true)@endverbatim
 *
 * Here is the related source code, which is taken from
 * <a href="https://github.com/TUC-ProAut/ros_pcdfilter/blob/master/src/pcdfilter_pa_node.cpp">pcdfilter_pa</a>:
 * @verbatim
cParameterPaRos paramloader;

// ...
paramloader.load("~/tf_lookup_time"   , filters_.tf_lookup_time_     );
paramloader.load("~/buffer_pointcloud", rosparams_.buffer_pointcloud_);
paramloader.load("~/debugging"        , rosparams_.debugging_        );@endverbatim
 *
 *
 * @section links_sec Links and packages
 *
 * Source code at github:
 *  + https://github.com/TUC-ProAut/ros_parameter
 *
 * ROS packages:
 *  + ros-indigo-parameter-pa
 *  + ros-kinetic-parameter-pa
 *  + ros-lunar-parameter-pa
 *
 *
 * @section doc_sec ROS Documentation
 *
 * ROS-Distribution    | Documentation
 * --------------------|---------------
 * Indigo              | <a href="http://docs.ros.org/indigo/api/parameter_pa/html/index.html">docs.ros.org</a>
 * Jade (EOL May 2017) | <a href="http://docs.ros.org/jade/api/parameter_pa/html/index.html">docs.ros.org</a>
 * Kinetic             | <a href="http://docs.ros.org/kinetic/api/parameter_pa/html/index.html">docs.ros.org</a>
 * Lunar               | <a href="http://docs.ros.org/lunar/api/parameter_pa/html/index.html">docs.ros.org</a>
 **/
